### PR TITLE
Implement Swift's physical lowering pass in the managed type system

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Implement Swift's physical lowering pass by re-using the interval traversal algorithm used in the explicit layout validator.

This does not hook it up to the JIT interface yet.

For the JIT interface API, I was thinking something like: `BOOL getSwiftLowering(CORINFO_CLASS_HANDLE clsHnd, uint8_t numLoweredElements, CORINFO_TYPE loweredElements[4]);`

This function would return `FALSE` for any types that should be passed by ref, and the `loweredElements` array would represent the equivalent primitive sequence (ending with `CORINFO_TYPE_EMPTY` if there are less than 4 primitives).

### CODE REVIEW

**Comment by jakobbotsch** [comment]
> For the JIT interface API, I was thinking something like: `BOOL getSwiftLowering(CORINFO_CLASS_HANDLE clsHnd, uint8_t numLoweredElements, CORINFO_TYPE loweredElements[4]);`

`bool` as a return value sounds like the API can fail. I would model it after `getSystemVAmd64PassStructInRegisterDescriptor`, e.g. something like
```
struct CORINFO_SWIFT_LOWERING
{
  bool passByReference;
  CORINFO_ELEMENT_TYPE loweredElements[4];
  size_t numLoweredElements;
};

void getSwiftTypeLowering(CORINFO_CLASS_HANDLE clsHnd, CORINFO_SWIFT_LOWERING* result)
```

File: src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/SwiftPhysicalLowering.cs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        public abstract record Lowering
+        {
+            private Lowering() { }
+
+            public sealed record InvalidType : Lowering;
+
+            public sealed record PassByRef : Lowering;
+
+            public sealed record PrimitiveSequence(ImmutableArray<CorInfoType> Sequence) : Lowering
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/SwiftPhysicalLowering.cs:23]: What are we getting out of these being `record`? We don't count every byte of the compiler size, but `record` is a death by a thousand papercuts and I don't think we use it anywhere in the compiler.

Comment (jkoritzinsky) [reply src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/SwiftPhysicalLowering.cs:23]: I personally like using records to model a discriminated-union-style structure. I'll change this to directly use the recommended JITInterface type.

**Comment by jakobbotsch** [comment]
I hit an assert in the lowering for this type:
```csharp
struct F5_S1_S0
{
    public short F0;
}

struct F5_S1
{
    public ulong F0;
    public long F1;
    public F5_S1_S0 F2;
    public long F3;
}
```

File: src/coreclr/tools/aot/ILCompiler.Compiler.Tests/ILCompiler.Compiler.Tests.Assets/SwiftTypes.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+[ExpectedLowering(ExpectedLoweringAttribute.Lowered.Int64, ExpectedLoweringAttribute.Lowered.Int64, ExpectedLoweringAttribute.Lowered.Int32, ExpectedLoweringAttribute.Lowered.Int64)]
+public struct F5_S2
+{
+    public ulong F0;
+    public long F1;
+    public F5_S2_S0 F2;
+    public long F3;
+}
```
Comment (jakobbotsch) [inline src/coreclr/tools/aot/ILCompiler.Compiler.Tests/ILCompiler.Compiler.Tests.Assets/SwiftTypes.cs:58]: I assume the last `Int64` here starts at offset 24, yet the primitives before it sum only to 20 in size. What is the right way for the JIT to figure out what part of the struct should corresponds to which specific primitive in the sequence?
It seems like the JIT-EE API should also include offsets for each primitive.

Comment (jkoritzinsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler.Tests/ILCompiler.Compiler.Tests.Assets/SwiftTypes.cs:59]: The JIT should assume natural alignment for each element in the primitive sequence. Any empty space in the middle of the struct due to alignment is unused padding in the struct that can be ignored.

That's how the generated code seems to look for the cases I've inspected.

**Review by jakobbotsch — APPROVED** [review]
LGTM, but you may want to get approval from someone more well-versed in the managed type system.

The support here is also for crossgen2, right? It would be good to update the PR title to reflect that if so.

File: src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                }
+
+                if (fieldType.Category is TypeFlags.UInt64 or TypeFlags.Int64)
+                {
+                    return LoweredType.Int64;
+                }
+
+                Debug.Assert(PointerSize == 8, "Swift interop is only supported on 64-bit platforms.");
+
+                if (fieldType.Category is TypeFlags.IntPtr or TypeFlags.UIntPtr || fieldType.IsPointer || fieldType.IsFunctionPointer)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs:82]:
```suggestion
                if (fieldType.Category is TypeFlags.IntPtr or TypeFlags.UIntPtr or TypeFlags.Pointer or TypeFlags.FunctionPointer)
```

**Review by MichalStrehovsky — APPROVED** [review]
Only looked at how the type system is used. Not reviewing the lowering logic or the correctness of ExplicitLayoutValidator refactor.

File: src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+                return new() { byReference = true };
+            }
+
+            LoweringVisitor visitor = new(type.Context.Target.PointerSize);
+            visitor.AddFields(type, addTrailingEmptyInterval: false);
+
+            List<CorInfoType> loweredTypes = visitor.GetLoweredTypeSequence();
+
+            // If a type has a primitive sequence with more than 4 elements, Swift passes it by reference.
+            if (loweredTypes.Count > 4)
```
Comment (kotlarmilos) [inline src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs:166]: Does it make sense to implement an early check in the `GetLoweredTypeSequence` and avoid lowering the full sequence if `loweredTypes.Count > 4`?

**Review by kotlarmilos — APPROVED** [review]
LGTM! I've left one comment for consideration. How will the JIT use it? Is it possible to share this implementation with the Mono runtime so that there is a unified implementation?

**Comment by jakobbotsch** [comment]
The algorithm lowers
```csharp
struct F2087_S0_S0
{
	public double F0;
	public float F1;
}

struct F2087_S0
{
	public F2087_S0_S0 F0;
	public int F1;
	public sbyte F2;
}
```
to
```
    [0] double
    [1] float
    [2] long
    [3] byte
```
`F2087_S0` is a 24 byte sized struct, but the returned lowering implies that `[3] byte` starts at offset 24 in the struct. Is it a bug in the lowering?

**Comment by jakobbotsch** [comment]
```csharp
struct F114_S0
{
	public float F0;
	public ushort F1;
	public short F2;
	public ushort F3;
}
```
seems to be lowered to
```
  Argument 0 of type SwiftAbiStress+F114_S0 must be passed as 2 primitive(s)
    [0] float
    [1] long
```
however, Swift seems to lower it to `float, i32, i16`: https://godbolt.org/z/WrE8ETzcb

**Comment by jkoritzinsky** [comment]
> How will the JIT use it? Is it possible to share this implementation with the Mono runtime so that there is a unified implementation?

The JIT will use this through the EE-JIT interface. We are going to need a separate implementation for CoreCLR (we can't use the managed type system there), so we can try to keep it to two implementations instead of three, but I'm not sure how easy that will be.

File: src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs (hunk 4 above)
Comment (jkoritzinsky) [reply src/coreclr/tools/Common/JitInterface/SwiftPhysicalLowering.cs:166]: I think it's fine. The number of frozen structs that would surpass the max size and be passed by-reference is likely very small (as if the type is being passed by-ref, it's not getting most of the benefits that being frozen allows).

**Comment by jkoritzinsky** [comment]
@jakobbotsch I found the issue. My implementation of the algorithm wasn't taking into account that we need to match padding of the original struct type. I've adjusted the algorithm and am pushing out a fix.

**Comment by jkoritzinsky** [comment]
Timeout is unrelated, merging in.

**Pull request merged**